wild/src/main.rs

```diff
@@ -1,4 +1,104 @@
+        } else {
+            true
+        }
+    });
+
+    match no_fork_subprocess {
+        false => {
+            let mut fds: [c_int; 2] = [0; 2];
+            match make_pipe(&mut fds) {
+                Ok(_) => {
+                    match unsafe { fork() } {
+                        0 => {
+                            // Fork success in child - Run linker in this process with remaining args
+                            let done_closure =
+                                move |exit_status: i32| inform_parent_done(&fds, exit_status);
+                            wild_lib::Linker::from_args(args.into_iter())?
+                                .run(Some(Box::new(done_closure)))
+                        }
+                        -1 => {
+                            // Fork failure in the parent - Fallback to running linker in this process
+                            // Err(anyhow!("Failed to fork"))
+                            wild_lib::Linker::from_args(args.into_iter())?.run(None)
+                        }
+                        _ => {
+                            // Fork success in the parent - wait for the child to "signal" us it's done
+                            let exit_status = wait_for_child_done(&fds)?;
+                            process::exit(exit_status);
+                        }
+                    }
+                }
+                Err(_e) => {
+                    // TODO do we want to log the error, or output a warning?
+                    // Err(anyhow!("Could not create named pipe: '{e}'"))
+                    // Failure to creat named pipe - Fallback to running linker in this process
+                    wild_lib::Linker::from_args(args.into_iter())?.run(None)
+                }
+            }
+        }
+        true => {
+            // Create a linker with remaining args and run it in this process
+            wild_lib::Linker::from_args(args.into_iter())?.run(None)
+        }
+    }
+}
+
+/// Inform the parent process that work of linker is done, sending the exit status over the pipe
+fn inform_parent_done(fds: &[c_int], exit_status: i32) {
+    unsafe {
+        libc::close(fds[0]);
+        let stream = libc::fdopen(fds[1], "w".as_ptr() as *const i8);
+        let bytes: [u8; 4] = exit_status.to_ne_bytes();
+        libc::fwrite(bytes.as_ptr() as *const c_void, 4, 1, stream);
+        libc::fclose(stream);
+    }
+}
+
+/// Wait for the child process to signal it is done, by returning an exit code on the pipe
+/// or for its unexpected death by closure of the pipe before receiving anything back
+fn wait_for_child_done(fds: &[c_int]) -> wild_lib::error::Result<i32> {
+    unsafe {
+        // close our sending end of the pipe
+        libc::close(fds[1]);
+        // open the other end of the pipe for reading
+        let stream = libc::fdopen(fds[0], "r".as_ptr() as *const i8);
+
+        // Wait for child to send exit_status or pipe to be closed
+        let mut response: [u8; 4] = [0u8; 4];
```

Comment:
Are you interested in the possibility of passing non-zero exit_status?
I don't know the linker code to know of other exit paths (couldn't see any!) that might return a non-zero/

We could remove code, by just waiting for pipe close, and closing from child...

Reply:
The pipe being closed could be because the subprocess is in the process of crashing, so I think we need to write something to the pipe, but it probably should be an exit status, since we should only do so when our exit status is success. Perhaps just write "OK" or something. The error-path should just be that the subprocess runs to completion and exits with a non-zero exit status. When the parent sees that the pipe was closed without any content, it should wait for the child to exit, which should happen almost immediately, then exit with the same exit code as the child.

Reply:
So I guess it's odd at the moment that there's support for sending a status code from the parent to the child, but the child only ever sends a status code of 0. If the child fails for some reason, then it closes the pipe and the parent then exits with -1 regardless of the failure mode of the child process. That said, I'm happy with the state that the PR is in if you'd just like to merge it as-is.